### PR TITLE
Feature/non blocking process io

### PR DESCRIPTION
**Pull request by vemilyus**
Hello there,

this is quite a big one, but that's a given considering what I did.

TL;DR: No more manual creation of threads, fewer threads used overall, cleaner async infrastructure

Rationale:
One thing that always bugged me were all the threads that were created all over the place. In my opinion this made the code harder to maintain and get into to start with. Also, Java's Process API with the streams that require their own thread each to pull the data is quite obsolete and a relic from a different time.

What I did:
1. Now using ForkJoinPool to execute FFmpeg and FFprobe tasks asynchronously, also helper threads for input and output are children of the returned Future and don't need to be cleaned up separately
2. Replaced Java Processes with NuProcess that handles input and output in a non-blocking fashion, removing the need to keep extra threads around
3. Adapted the output parsers to work non-blocking because they are now controlled by NuProcess
4. They were already very good in that regard, didn't have to do much! 🥇 

Basically the process hierarchy now looks like this:
```
ProcessFuture -> ForkJoinTask -> Helpers
                              -> NuProcess -> Output parsers
```

NuProcess manages all its own threads and doesn't expose them to the world, for more info please see the [NuProcess](https://github.com/brettwooldridge/NuProcess) page.

5. The return type of the async executions is now a proper Java Future which also gives access to the `ProcessAccess`, the interface you use to stop a process. Or you can just use `cancel`.
6. Process and output listeners can no longer stop the process. 💭 To make this possible with the new process infrastructure would have created a whole new mess, also I don't think that observers should be able to influence the operation of the dispatcher.

All in all I did very little to change the actual functionality of handling FFmpeg output, there everything stays the same. If it ain't broke don't fix it. I just replaced the glue holding it all together, it was already showing cracks.

Apart from the `executeAsync` methods and `FFprobe.setFormatParser` I didn't change anything at the API level for `FFmpeg` and `FFprobe`.

Please look it over and don't hesitate to hit me up with questions or concerns 😉

### CODE REVIEW

**Comment by kokorin** [comment]
Hello @vemilyus 
Thanks for this PR, it's an interesting idea to use `ForkJoinPool` to run all required threads.

It's hard to analyze this PR because of its size. It would be easier for me to accept several smaller PRs. 
I noticed you added 2 Maven dependencies it would be great if this project stay with minimal dependencies, i.e. only slf4j-api.

Why is it required to change ffprobe result parser?
Also project version should not be changed, version in develop branch  already set to the next release version.
If you change a file you should not remove an original author from a copyright notice.

So, please minimize PR size, change only  `ForkJoinPool` related code.

Best regards,
Denis Kokorin.

**Review by kokorin — CHANGES_REQUESTED** [review]
Too many changes

**Comment by vemilyus** [comment]
Hi @kokorin,

please allow to break down your concerns and address them one at a time:

> It's hard to analyze this PR because of its size. It would be easier for me to accept several smaller PRs.

This is partly true. But the changes are systemic and depend on each other.

> I noticed you added 2 Maven dependencies it would be great if this project stay with minimal dependencies, i.e. only slf4j-api.

A worthy goal to be sure, but a balance must be struck between functionality and paradigm. I agree that `annotations-java5` is most likely not necessary (it's mostly a development aid) but the entire PR is built on the foundation of `nuprocess` which is there to effectively change the way external processes are launched and handled, and the basis of the PR.

> Why is it required to change ffprobe result parser?

This is only a minor change. The only change is how the lines are supplied, before they were retrieved from an `InputStream` in a loop, now the separate lines are supplied by the new process infrastructure one at a time. The rest of the functionality stays the same.

> Also project version should not be changed, version in develop branch already set to the next release version.

This is fair. I can change this to the appropriate value.

> If you change a file you should not remove an original author from a copyright notice.

An oversight on my part. I will fix this!

> So, please minimize PR size, change only `ForkJoinPool` related code.

That would defeat the purpose of this PR. `ForkJoinPool` related code is only the last piece of the puzzle, not the defining feature. The core of this PR lies in the use of `NuProcess` and the changes related to that.

I understand the PR looks unwieldy, but I recommend you check out the branch and look at the changes in your IDE and not just here on GitHub. I think then most things should become very clear and easily digestible. One objective of the PR is to reduce cognitive load by removing all that manual thread-management code and using a battle-tested alternative to the Process API to make this possible.

Please let me know if this addresses your concerns in an adequate fashion or you have other gripes with what I created.

Best regards,
Alex Katlein

**Comment by jonfryd** [comment]
Awesome, @vemilyus! I was actually just about to make a feature request for async probing when I noticed your PR.

**Comment by vemilyus** [comment]
Hi @kokorin,

I'm sad to see that there has been no further activity on this PR.
While I understand some of your points I'm nonetheless disappointed that this contribution is seemingly going nowhere.

Thanks for all the work you did to even get the library to this point, but as it stands now we're at an impasse.
Closing the PR!
